Turn into a buildable gem and add a simple CLI

mhl opened the pull request:
I know this isn't an ideal PR of the "one small well-defined change" type that we like, so making this is more asking the question, "Is it OK to wrap the functionality of `history.rb` within this repo in this way?" (It's also a bit undertested at the moment because of the tight deadline for when this needs to be ready.)

I think the `WikidataPositionHistory::PageRewriter` class is all that I should need in the new repository I'm working on, which will have the Sinatra app and cron job wrappers for the WMF tools server.

I've tested the CLI tool to try the error conditions and to check it could rewrite pages successfully - this seems to be fine: https://www.wikidata.org/wiki/User:Mhl20/Prime_minister_test

... with the caveat that at the moment it will fail to update if someone does `{{PositionHolderHistory|blah}}` for example.

mhl committed:
Add basic gem configuration from 'bundle gem --mit --test=minitest ...'

mhl committed:
Explicitly just require rest-client

Previously this was being brought in with "require 'wikidata/fetcher'"
but the wikidata-fetcher gem is no actually required, so we're using
more minimal dependencies.

mhl committed:
Remove a runtime dependency on pry

mhl committed:
Turn the history.rb script into a wikitext_history function

We will want to invoke this code to generate the wikitext representing
the history of the people who have held a particular position from Ruby.

mhl committed:
Ignore a Rubocop cache file

mhl committed:
Add a high level test of wikitext_history's functionality

mhl committed:
Mock the HTTP request to the Wikidata Query Service

mhl committed:
Add the initializer and attributes of the PageRewriter class

mhl committed:
Add a runtime dependency on mediawiki-page-replaceable_content 0.1.2

We'll need this for rewriting pages.

mhl committed:
Add a CLI for wikidata_position_history

This can be invoked, for example as:

  bundle exec exe/wikidata_position_history 'mediawiki_site=www.wikidata.org&page_title=User%3AMhl20%2FPrime_minister_test'

... to update the output after the PositionHolderHistory template tag
on the User:Mhl20/Prime_minister test page on https://www.wikidata.org

The exe/wikidata_position_history command line wrapper is essentially
copied from:

  https://github.com/everypolitician/compare_with_wikidata/blob/master/exe/compare_with_wikidata

tmtmtmtm commented on LICENSE.txt:
```diff
@@ -0,0 +1,21 @@
+The MIT License (MIT)
+
+Copyright (c) 2017 Mark Longair
```
Is this meant to be mySociety?

mhl replied:
Yes, sorry - that's just what `bundle gem` defaulted to.

tmtmtmtm commented on README.md:
```diff
@@ -0,0 +1,39 @@
+
+Welcome to your new gem! In this directory, you'll find the files you need to be able to package up your Ruby library into a gem. Put your Ruby code in the file `lib/wikidata_position_history`. To experiment with that code, run `bin/console` for an interactive prompt.
+
+TODO: Delete this and the text above, and describe your gem
```
Looks like you forgot to edit this file (here, and a few places below)

mhl replied:
I've updated that file with basic usage instructions, etc.

tmtmtmtm commented on history.rb:
```diff
@@ -3,7 +3,7 @@
 # frozen_string_literal: true
```
NB: there's a typo in the commit message here "Explicity" → "Explicitly" (and "no actually" → "not actually")

tmtmtmtm commented on Gemfile:
```diff
@@ -1,29 +1,6 @@
-gem 'wikisnakker', github: 'everypolitician/wikisnakker'
-
-group :quality do
-  gem 'rubocop'
```
You seem to have removed the `rubocop` dependency (even though you've also committed an extra rubocop-related file). I think it's worth keeping it as a dev dependency

mhl replied:
Oops, yes, that was unintentional.

tmtmtmtm commented on wikidata_position_history.gemspec:
```diff
@@ -0,0 +1,40 @@
+Gem::Specification.new do |spec|
+  spec.name          = "wikidata_position_history"
+  spec.version       = WikidataPositionHistory::VERSION
+  spec.authors       = ["Tony Bowden"]
```
I think you've done enough here that you should either add yourself to this, or explicitly make it the team instead of an individual

tmtmtmtm commented on wikidata_position_history.gemspec:
```diff
@@ -0,0 +1,40 @@
+  spec.add_development_dependency "rake", "~> 10.0"
+  spec.add_development_dependency "minitest", "~> 5.0"
+  spec.add_development_dependency 'pry', '~> 0.10'
+  spec.add_development_dependency 'webmock', '~> 3.0.0'
```
The mismatched quote style here is a little jarring. I think it's worth running everything through `rubocop -a` — it's OK to do that as a commit of its own at the end if you don't want to rewrite each existing commit.

mhl replied:
I've fixed the gemspec by rewriting the history, but the other changes it suggests I've added in a new commit.

tmtmtmtm reviewed (commented):
A couple of nits in-place, but generally I'd say feel free to commit at-will, and we can refactor later when required.

The one thing I'd quite like to see, though, is there still being a built-in way to replicate the existing behaviour — i.e. a script that can be passed a pageid, and which spits out wikitext suitable for manual pasting into a wikipage (i.e. which doesn't actually do the rewriting itself).  

The only other (very minor) concern I have at the minute is that the public interface of WikidataPositionHistory is very broad — there are a lot of things in there that I suspect should really be private. However, as it's very very unlikely that anyone but us is going to use this at all any time soon, never mind box us in to supporting a specific interface, I wouldn't worry too much about that right now.

mhl committed:
Apply (most of) the corrections made by rubocop -a

mhl committed:
Minimize the public interface of PageRewriter

Very few of the methods of WikidataPositionHistory::PageRewriter
actually need to be public, so make them private instead.

mhl committed:
Add a script with the original behaviour (item ID -> wikitext)

This preserves the original behaviour of the history.rb script, which is
still useful.

mhl commented:
I think I've addressed the inline comments - I've pushed a rewritten version of the branch, since otherwise the fixup commits were getting very tangled. Also:

> The one thing I'd quite like to see, though, is there still being a built-in way to replicate the existing behaviour — i.e. a script that can be passed a pageid, and which spits out wikitext suitable for manual pasting into a wikipage (i.e. which doesn't actually do the rewriting itself).

I've added a `exe/position-history-for-item` script that does that.

> The only other (very minor) concern I have at the minute is that the public interface of WikidataPositionHistory is very broad — there are a lot of things in there that I suspect should really be private. However, as it's very very unlikely that anyone but us is going to use this at all any time soon, never mind box us in to supporting a specific interface, I wouldn't worry too much about that right now.

Yes, good point - I've change that so most of the methods are now private.
